### PR TITLE
Simplify Field Caps logic further

### PR DESCRIPTION
**Pull request by original-brownbear**
Lots of memory savings here. Stop using streams in some spots, remove intermediate collections of stream to list where unnecessary and some other obvious cleanups.

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/CCSFieldCapabilitiesIT.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public void testFailuresFromRemote() {
             .setIndexFilter(new ExceptionOnRewriteQueryBuilder())
             .get();
         assertThat(response.getIndices()[0], equalTo(localIndex));
-        assertThat(response.getFailedIndices()[0], equalTo("remote_cluster:*"));
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/CCSFieldCapabilitiesIT.java:69]: We never used the actual string values for this thing in production, only the count so I made it count only.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilities.java [hunk 2]
```diff
@@ -473,10 +472,6 @@ public String toString() {
         return Strings.toString(this);
     }
 
-    static FieldCapabilities buildBasic(String field, String type, String[] indices) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilities.java:476]: Single use, I inlined this.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java [hunk 3]
```diff
@@ -58,11 +58,8 @@ FieldCapabilitiesIndexResponse fetch(
     ) throws IOException {
         final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         final IndexShard indexShard = indexService.getShard(shardId.getId());
-        if (alwaysMatches(indexFilter)) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java:61]: Just shortening the code a little here

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java [hunk 4]
```diff
@@ -81,8 +81,8 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeMap(responseMap, StreamOutput::writeString, (valueOut, fc) -> fc.writeTo(valueOut));
         }
 
-        List<FieldCapabilitiesIndexResponse> getResponses() {
-            return indices.stream().map(index -> new FieldCapabilitiesIndexResponse(index, indexMappingHash, responseMap, true)).toList();
+        Stream<FieldCapabilitiesIndexResponse> getResponses() {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:84]: No need to collect this if we turn it into a stream in the only callsite right away anyway.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesNodeRequest.java [hunk 5]
```diff
@@ -143,12 +143,16 @@ public ActionRequestValidationException validate() {
     public String getDescription() {
         final StringBuilder stringBuilder = new StringBuilder("shards[");
         Strings.collectionToDelimitedStringWithLimit(shardIds, ",", "", "", 1024, stringBuilder);
+        return completeDescription(stringBuilder, fields, filters, allowedTypes);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesNodeRequest.java:146]: I optimised the string building here a little and dried this up. Not sure this ever gets hot but no need to burn more heap than necessary when building this large string.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java [hunk 6]
```diff
@@ -322,51 +322,55 @@ private static Map<String, Map<String, FieldCapabilities>> buildResponseMap(
...
-            Optional<Function<Boolean, FieldCapabilities>> unmapped = Optional.empty();
+            Function<Boolean, FieldCapabilities> unmapped = null;
             if (includeUnmapped) {
                 // do this directly, rather than using the builder, to save creating a whole lot of objects we don't need
                 unmapped = getUnmappedFields(
-                    indexResponsesMap.keySet(),
+                    indices,
                     entry.getKey(),
-                    typeMapBuilder.values().stream().flatMap(FieldCapabilities.Builder::getIndices).collect(Collectors.toSet())
+                    typeMapBuilder.stream().flatMap(t -> t.getValue().getIndices()).collect(Collectors.toSet())
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:336]: This is better now but still nasty, doing this kind of stream allocation in a hot loop isn't great.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java [hunk 7]
```diff
@@ -322,51 +322,55 @@ private static Map<String, Map<String, FieldCapabilities>> buildResponseMap(
...
-                            .stream()
-                            .map(e -> Map.<String, Function<Boolean, FieldCapabilities>>entry(e.getKey(), e.getValue()::build)),
-                        unmapped.stream().map(f -> Map.entry("unmapped", f))
-                    ).collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().apply(multiTypes)))
-                )
-            );
+            final int resSize = typeMapBuilder.size() + (unmapped == null ? 0 : 1);
+            boolean multiTypes = resSize > 1;
+            final Map<String, FieldCapabilities> res = Maps.newHashMapWithExpectedSize(resSize);
+            for (Map.Entry<String, FieldCapabilities.Builder> e : typeMapBuilder) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:343]: Building to a pre-sized map outright without all the endless intermediary objects should save quite a bit of heap here.

File: server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/CCSFieldCapabilitiesIT.java (hunk 1 above)
Comment (romseygeek) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/CCSFieldCapabilitiesIT.java:69]: Does this change the REST response?

Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/CCSFieldCapabilitiesIT.java:69]: No, the method is only used to get the array length in production, it's not used in serialisation.

**Review by romseygeek — APPROVED** [review]
LGTM!

**Comment by original-brownbear** [comment]
Thanks Alan!

**Pull request merged**